Support for readdir_buffer option

rovo89 opened the pull request.

FGasper commented on LibNFS.xs:
```diff
@@ -983,6 +983,33 @@ static void _set_unix_authn(pTHX_ struct nfs_context* nfs, SV* value_sv) {
+        uint32_t nums_count = 1 + av_len(av);
+        if (nums_count != 2) croak("“%s” must contain exactly 2 numbers", "readdir_buffer");
+
+        uint32_t nums[nums_count];
```
Might as well hard-code this at 2 in order to avoid a variadic array. Maybe create a #define for it, e.g.:
```
#define MAX_BUFFER_SIZE_ARRAY_SIZE 2
```
… then:
```
    UV nums[MAX_BUFFER_SIZE_ARRAY_SIZE];
```

rovo89 replied:
I had it hard-coded first - then thought if there's any need to allow a different number of arguments for some reason (or the code gets copied), it should react dynamically. I can change it though if you like.

FGasper replied:
AFAICT, as the code stands there would still need to be changes if there were a case for something besides 2 arguments.

rovo89 replied:
Done

FGasper commented on LibNFS.xs:
```diff
@@ -983,6 +983,33 @@ static void _set_unix_authn(pTHX_ struct nfs_context* nfs, SV* value_sv) {
+            SV** svp = av_fetch(av, n, 0);
+            assert(svp);
+
+            nums[n] = exs_SvUV(*svp);
```
Should reject if `exs_SvUV(*svp) > UINT32_MAX` before assigning into `nums[n]`.

rovo89 replied:
Same here. Will adjust this in _set_unix_authn() and use uint64_t here as upstream. Do I still need a check then?

FGasper replied:
I wouldn’t say `_set_unix_authn()` needs a change here, though if there’s a disparity in int sizes there then that should be addressed separately.

If upstream takes a u32, then we need to ensure that our UV doesn’t exceed what a u32 can hold.

rovo89 replied:
I have done various tests with UID 1_000_000_000, 4_000_000_000 (> INT32_MAX but < UINT32_MAX) and 5_000_000_000 (> UINT32_MAX). While the first two worked fine and survived the implicit casts back and forth, 5_000_000_000 appeared as 705_032_704, i.e. only the lower 4 bytes were considered. So I created #8 to restrict the values to UINT32_MAX.

rovo89 replied:
> Should reject if `exs_SvUV(*svp) > UINT32_MAX` before assigning into `nums[n]`.

Done.

rovo89 replied:
While I was checking in Wireshark, I noticed that the field is also just 4 bytes. I somehow misinterpreted the type definitions. I'll send a PR for libnfs and will also ask about nfs_set_uid().

FGasper commented on LibNFS.xs:
```diff
@@ -983,6 +983,33 @@ static void _set_unix_authn(pTHX_ struct nfs_context* nfs, SV* value_sv) {
+
+        for (unsigned n=0; n<nums_count; n++) {
+            SV** svp = av_fetch(av, n, 0);
+            assert(svp);
```
NB: Perl black-holes `assert()`. It _should_ never fail, though, since you just verified the array’s length above.

rovo89 replied:
I just copied this from _set_unix_authn().

FGasper replied:
Yeah it’s fine; just FYI.

FGasper commented on LibNFS.xs:
```diff
@@ -983,6 +983,33 @@ static void _set_unix_authn(pTHX_ struct nfs_context* nfs, SV* value_sv) {
+            nums[n] = exs_SvUV(*svp);
+        }
+
+        nfs_set_readdir_max_buffer_size(nfs, nums[0], nums[1]);
```
Please refactor a bit so there’s only 1 call to this function.

rovo89 replied:
OK.

rovo89 replied:
Done.

FGasper commented on lib/Net/LibNFS.pm:
```diff
@@ -152,7 +152,7 @@ at once. Clobbers (and is clobbered by) C<uid> and C<gid>.
 =item * C<pagecache>, C<pagecache_ttl>, C<readahead>
 
-=item * C<readmax>, C<writemax>
+=item * C<readmax>, C<writemax>, C<readdir_buffer>
```
I wonder if `readdir_buffer_size` would be easier to grok. At first blush I’m not sure if this is a boolean or what.

rovo89 replied:
I used the shorter form upstream because it's faster to type in the URLs. I can change it here, but then it would differ from upstream. In any case this needs more docs to explain that either a scalar or array ref can be used.

FGasper replied:
> but then it would differ from upstream

How so? Upstream has `nfs_set_readdir_max_buffer_size`?

rovo89 replied:
Yes, that's the function name. However the example tools like `nfs-ls` take a URL, and there it's the shorter form:
https://github.com/sahlberg/libnfs/blob/4f7f645cd8d9a7c6175b6ce986c92e311a809bda/lib/libnfs.c#L338-L347

Because well, `nfs://some.host/the/export/path?readdir_max_buffer_size=65536` felt too clunky to type. There was a precedent for `nfs_set_interface()` where the option is named `if`.

Anyway, up to you, just let me know how the option should be named in the Perl module.

rovo89 commented:
Thanks for your review. I'll take care of these things when I'm back to work on Monday.

rovo89 commented:
I implemented the requested changes. Please check if `_set_readdir_max_buffer_size()` is fine now. I also added a sentence about the option in the README.

By the way, upstream has some new commits to fix Windows support and update workflows for BSD. Maybe that can help to fix the checks.

FGasper commented on LibNFS.xs:
```diff
@@ -983,6 +983,50 @@ static void _set_unix_authn(pTHX_ struct nfs_context* nfs, SV* value_sv) {
+    uint32_t dircount, maxcount;
+
+    if (SvROK(value_sv) && (SvTYPE(SvRV(value_sv)) == SVt_PVAV)) {
+        #define MAX_BUFFER_SIZE_ARRAY_SIZE 2
```
This should IMO be outside the function, but I can do that in a follow-up commit.

FGasper commented on LibNFS.xs:
```diff
@@ -983,6 +983,50 @@ static void _set_unix_authn(pTHX_ struct nfs_context* nfs, SV* value_sv) {
+
+        if (value > UINT32_MAX) {
+            croak("%s: value (%" UVuf ") exceeds maximum (%u)", "readdir_buffer", value, UINT32_MAX);
+        }
```
Should deduplicate; I’ll do in a follow-up.

rovo89 replied:
Maybe as a helper in easyxs?

FGasper commented on LibNFS.xs:
```diff
@@ -983,6 +983,50 @@ static void _set_unix_authn(pTHX_ struct nfs_context* nfs, SV* value_sv) {
+            assert(svp);
+
+            UV value = exs_SvUV(*svp);
+            if (value > UINT32_MAX) {
```
So, I looked at the source … these are actually defined as U64, not U32.

I’ll just create a follow-up PR to address this & other stuff.

rovo89 replied:
> these are actually defined as U64, not U32.

Yes, but that was my mistake. NFS protocol has only room for U32. See https://github.com/sahlberg/libnfs/pull/408.

FGasper replied:
Ahh. 👍
